leash.go

```diff
@@ -67,15 +74,36 @@ func run(options GlobalOptions) {
 		Options: options.Tail,
 	}
 	if options.TailSample {
-		linesChans, err = tail.GetSampledEntries(tc, options.SampleRate)
+		linesChans, err = tail.GetSampledEntries(tc, options.SampleRate, abort)
 	} else {
-		linesChans, err = tail.GetEntries(tc)
+		linesChans, err = tail.GetEntries(tc, abort)
 	}
 	if err != nil {
 		logrus.WithFields(logrus.Fields{"err": err}).Fatal(
 			"Error occurred while trying to tail logfile")
 	}
 
+	// set up our signal handler, now that we know how many files we're tailing,
+	// we can send the right number of abort signals.
+	go func() {
+		sig := <-sigs
+		fmt.Fprintf(os.Stderr, "Aborting! Caught signal \"%s\"\n", sig)
+		fmt.Fprintf(os.Stderr, "Cleaning up...\n")
+		close(abort)
+		// and if they insist, catch a second CTRL-C or timeout on 10sec
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		go func() {
+			select {
+			case <-sigs:
+				fmt.Fprintf(os.Stderr, "Caught second signal... Aborting.\n")
+				os.Exit(1)
+			case <-time.After(10 * time.Second):
+				fmt.Fprintf(os.Stderr, "Taking too long... Aborting.\n")
+				os.Exit(1)
+			}
+		}()
```

Comment:
aha!

ok. while calling a go func inside another go func isn't bad in and of itself, it **isn't** in fact necessary in this case!

your goal here was presumably to:
1. not block successful completion of this program by potentially waiting on a second signal or timeout
2. allow an exit valve if the "successful" completion takes too long after the first signal capture.

but! the fact that this is happening inside a goroutine in the first place means that #2 wouldn't happen, and since successful completion of this program means that everything in `linesChan` is closed and processed, the `exit 0` path is hit and so #1 is also irrelevant (we aren't forced, for example, to wait for the 10s timeout.)

so you can go straight from:

```go
close(abort)
select {
  case <-sigs:
    // ...
  case <-time.After(10 * time.Second):
    // ...
}
```

... and check all your boxes!